DotVVM.Diagnostics.StatusPage/DotHtmlFilesRuntimePrecompiler.cs

```diff
@@ -0,0 +1,180 @@
+                        {
+                            tempList.Add(new DotHtmlFileInfo()
+                            {
+                                VirtualPath = masterPage
+                            });
+                        }
+                    }
+
+                    file.Status = CompilationState.CompletedSuccessfully;
+                    file.Exception = null;
+                }
+                catch (Exception e)
+                {
+                    file.Status = CompilationState.CompilationFailed;
+                    file.Exception = e.Message;
+                }
+            }
+        }
+        /// <summary>
+        /// Builds all dothtml file registered in configuration. When the files are already compiled and cached, compilation is not executed.  
+        /// </summary>
+        public async Task<bool> CompileAll()
+        {
+            ResolveAllFiles();
+            var tempMasterPages = new ConcurrentBag<DotHtmlFileInfo>();
+            await BuildPagesAndControls(tempMasterPages);
+            await BuildMasterPages(tempMasterPages);
+            return MasterPages.Concat(Controls).Concat(Routes).All(s => s.Status != CompilationState.CompilationFailed);
+        }
+
+        private async Task<ConcurrentBag<DotHtmlFileInfo>> BuildMasterPages(ConcurrentBag<DotHtmlFileInfo> tempMasterPages)
+        {
+            while (tempMasterPages.Count > 0)
+            {
+                tempMasterPages.ToList().ForEach(masterPages.Add);
+                masterPages = new ConcurrentBag<DotHtmlFileInfo>(masterPages.Distinct());
+                tempMasterPages = new ConcurrentBag<DotHtmlFileInfo>();
+
+                if (statusPageOptions.BuildInParallel)
+                {
+                    await Task.WhenAll(masterPages.Select(i => Task.Run(() => BuildView(i, tempMasterPages))).ToArray());
+                }
+                else
+                {
+                    foreach (var item in masterPages)
+                    {
+                        BuildView(item, tempMasterPages);
+                    }
+                }
+            }
+
+            return tempMasterPages;
+        }
+
+        private async Task BuildPagesAndControls(ConcurrentBag<DotHtmlFileInfo> tempMasterPages)
+        {
+            if (statusPageOptions.BuildInParallel)
+            {
+                var compileTasks = Routes.Select(a => Task.Run(() => BuildView(a, tempMasterPages))).ToList();
+                compileTasks.AddRange(Controls.Select(a => Task.Run(() => BuildView(a, tempMasterPages))).ToList());
+                await Task.WhenAll(compileTasks.ToArray());
+            }
+            else
+            {
+                foreach (var item in Routes)
+                {
+                    BuildView(item, tempMasterPages);
+                }
+                foreach (var item in Controls)
+                {
+                    BuildView(item, tempMasterPages);
+                }
```

Comment:
Generally, it would be nicer to deduplicate the logic. For example, you could create a enumerable of `Action`s that perform the compilation. Then either execute them in sequence or in parallel.

Reply:
This would also mean that you can dedupe the parallel/sequence condition with `BuildMasterPages`